client.go

```diff
@@ -340,8 +347,18 @@ func (c *client) handleVersionNegotiationPacket(p *receivedPacket) {
 	c.logger.Infof("Received a Version Negotiation packet. Supported Versions: %s", hdr.SupportedVersions)
 	newVersion, ok := protocol.ChooseSupportedVersion(c.config.Versions, hdr.SupportedVersions)
 	if !ok {
-		c.session.destroy(fmt.Errorf("No compatible QUIC version found. We support %s, server offered %s", c.config.Versions, hdr.SupportedVersions))
-		c.logger.Debugf("No compatible QUIC version found.")
+		fail := func(e error) {
+			c.session.destroy(e)
+			c.logger.Debugf("No compatible QUIC version found.")
+		}
+		errImmediate := fmt.Errorf("No compatible QUIC version found. We support %s, server offered %s", c.config.Versions, hdr.SupportedVersions)
+		errTimeout := fmt.Errorf("Timed out: No compatible QUIC version found. We support %s, server offered %s", c.config.Versions, hdr.SupportedVersions)
+		sess, ok := c.session.(*session)
+		if !ok { // this catches MockQuicSession
```

Comment:
How hard would it be to make `MockQuicSession` implement `maybeRecover`?  Would that allow us to remove this conditional?

Reply:
`MockQuicSession` is [generated from](https://github.com/lucas-clemente/quic-go/blob/7df59b855f69adebbfaa95803a5976ac35cc49e0/mockgen.go#L17) `quicSession`, so you'd have to add `maybeRecover` to that interface. Generating the mocks can be done by running
```bash
GO111MODULE=off go generate .
```
I know it's messy to turn module support off for that, but mockgen is unfortunately not as smooth as I wish it would be, and it requires us to jump through a bunch of hoops to generate mocks for private methods (and even more for internal packages).